llama_index/node_parser/text/semantic_splitter.py

```diff
@@ -0,0 +1,198 @@
+from typing import Any, Callable, List, Optional, Sequence, TypedDict
+
+import numpy as np
+
+from llama_index.bridge.pydantic import Field
+from llama_index.callbacks.base import CallbackManager
+from llama_index.embeddings.base import BaseEmbedding
+from llama_index.node_parser import NodeParser
+from llama_index.node_parser.interface import NodeParser
+from llama_index.node_parser.node_utils import build_nodes_from_splits
+from llama_index.node_parser.text.utils import split_by_sentence_tokenizer
+from llama_index.schema import BaseNode, Document
+from llama_index.utils import get_tqdm_iterable
+
+DEFAULT_OG_TEXT_METADATA_KEY = "original_text"
+
+
+class SentenceCombination(TypedDict):
+    sentence: str
+    index: int
+    combined_sentence: str
+    combined_sentence_embedding: List[float]
+    distance_to_next: float
+
+
+class SemanticSplitterNodeParser(NodeParser):
```

Comment:
I wonder if we can add a unit test or two for this. Maybe by mocking the embedding model 🤔

Reply:
Or embeddings aside, we could refactor the `build_semantic_nodes_from_documents()` into a few different functions to make the overall logic more testable

Reply:
Yeah, I was wondering the same.  I actually just pinged you on Discord mentioning this :)

Reply:
ah, yeah, that's fair on splitting it up :+1: